### PR TITLE
#755 backend to supply autocomplete values

### PR DESCRIPTION
**Pull request by mspratt-biot**

### CODE REVIEW

**Comment by mspratt-biot** [comment]
Not too sure of the folder structure/where to place these new files. And before I went too far, I wanted to have some eyes on the direction I'm heading.

File: udmif/api/src/device/AutocompleteDataSource.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { GraphQLDataSource } from 'apollo-datasource-graphql/dist/GraphQLDataSource';
+import { AutocompleteOptions } from './model';
+import { AutocompleteDAO } from './dao/AutocompleteDAO';
+
+export class AutocompleteDataSource extends GraphQLDataSource {
```
Comment (neil-buildingsiot) [inline udmif/api/src/device/AutocompleteDataSource.ts:5]: This should not extend GraphQLDataSource. It is not fetching data from a graphql source.

File: udmif/api/src/device/dao/mongodb/MongoAutocompleteDAO.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  constructor(private db: Db) {}
+
+  async getAutocompleteSuggestions(autocompleteOptions: AutocompleteOptions): Promise<string[]> {
+    const { entity, field, term, limit } = autocompleteOptions;
+
+    try {
+      const results = await this.db.collection(entity).distinct(field);
+
+      if (!term) {
+        return results.slice(0, limit);
```
Comment (neil-buildingsiot) [inline udmif/api/src/device/dao/mongodb/MongoAutocompleteDAO.ts:15]: Can we pass this limit as part of the original query above?

File: udmif/api/src/device/schema.graphql [hunk 3]
```diff
@@ -89,4 +89,4 @@ type Query {
   points( deviceId: ID!) : [Point!]!
   # Get the auto-complete search field suggestions
   autocompleteSuggestions(autocompleteOptions: AutocompleteOptions!): [String]
```
[the remark this answers is not in the document]
Comment (mspratt-biot) [reply udmif/api/src/device/schema.graphql:91]: You're right.

File: udmif/api/src/device/dao/mongodb/MongoAutocompleteDAO.ts (hunk 2 above)
Comment (mspratt-biot) [reply udmif/api/src/device/dao/mongodb/MongoAutocompleteDAO.ts:15]: So i was just following how it was done for devices, where we pass in a single object called `searchOptions` and `limit` is also in there. Should devices change too then?

File: udmif/api/src/device/AutocompleteDataSource.ts (hunk 1 above)
Comment (mspratt-biot) [reply udmif/api/src/device/AutocompleteDataSource.ts:5]: Hmm ok, I'll discuss with Step. I also didn't know really where to put this file. I just wanted to make something at least work.

Comment (neil-buildingsiot) [reply udmif/api/src/device/AutocompleteDataSource.ts:5]: Oh, yeah i didnt notice the file loc. I dont think we should have it under device, but we can discuss with Steph when he's back.

File: udmif/api/src/device/dao/mongodb/MongoAutocompleteDAO.ts (hunk 2 above)
Comment (neil-buildingsiot) [reply udmif/api/src/device/dao/mongodb/MongoAutocompleteDAO.ts:15]: Will need to chat about this. I'm just trying to avoid overfetching.

**Pull request closed without merging**